File: python3/stats.py
from math import isfinite, isnan, nan, sqrt


class SampleStatistics(object):

    def __init__(self) -> None:
        self.max = nan
        self.min = nan
        self.samples: int = 0
        self.sum: float = 0
        self.sum_of_squares: float = 0

    def put(self, v: float) -> None:
        if self.samples == 0:
            self.min = v
            self.max = v
        elif self.min > v:
            self.min = v
        elif self.max < v:
            self.max = v
        self.samples += 1
        self.sum += v
        self.sum_of_squares += v * v

    def mean(self) -> float:
        if self.samples > 0 and isfinite(self.sum):
            r = self.sum / self.samples
            return max(self.min, min(self.max, r))
        else:
            return nan

    def variance(self) -> float:
        if self.samples > 1 and isfinite(self.sum_of_squares):
            n = self.samples
            r = (self.sum_of_squares - self.sum * self.sum / n) / (n - 1)
            return max(0, r)
        else:
            return nan

    def deviation(self) -> float:
        r = sqrt(self.variance())
        if isnan(r):
            return nan
        else:
            return min(self.max - self.min, r)

File: go/src/main/stats.go
package main

import (
	"math"
)

type SampleStatistics struct {
	max          float64
	min          float64
	samples      int
	sum          float64
	sumOfSquares float64
}

func (s *SampleStatistics) Put(v float64) {
	if s.samples == 0 {
		s.min = v
		s.max = v
	} else if s.min > v {
		s.min = v
	} else if s.max < v {
		s.max = v
	}
	s.samples++
	s.sum += v
	s.sumOfSquares += v * v
}

func (s *SampleStatistics) Max() float64 {
	return s.max
}

func (s *SampleStatistics) Min() float64 {
	return s.min
}

func (s *SampleStatistics) Mean() float64 {
	r := s.sum / float64(s.samples)
	if r < s.min {
		return s.min
	}
	if r > s.max {
		return s.max
	}
	return r
}

func (s *SampleStatistics) Variance() float64 {
	n := float64(s.samples)
	r := (s.sumOfSquares - s.sum*s.sum/n) / (n - 1)
	if r < 0 {
		return 0
	}
	return r
}

func (s *SampleStatistics) Deviation() float64 {
	r := math.Sqrt(s.Variance())
	m := s.max - s.min
	if r > m {
		return m
	}
	return r

}
